compiler/circle-mpqsolver/src/bisection/ErrorApproximator.cpp

```diff
@@ -0,0 +1,386 @@
+    }
+    break;
+    case CircleOpcode::FULLY_CONNECTED:
+    {
+      auto fc = loco::must_cast<const CircleFullyConnected *>(node);
+      weight = fc->weights();
+    }
+    break;
+    default:
+      break;
+  }
+
+  return weight;
+}
+
+void iterate_per_channel(const CircleNode *node, IterFunc func)
+{
+  CircleConst *weight = loco::must_cast<CircleConst *>(get_weight(node));
+  loco::TensorShape dimension;
+  set_weight_dim(node, weight, dimension);
+  uint32_t indices[4] = {
+    0,
+  };
+
+  auto channel_dim_index = get_channel_dim_index(node);
+
+  for (indices[0] = 0; indices[0] < dimension.dim(0).value(); indices[0]++)
+  {
+    for (indices[1] = 0; indices[1] < dimension.dim(1).value(); indices[1]++)
+    {
+      for (indices[2] = 0; indices[2] < dimension.dim(2).value(); indices[2]++)
+      {
+        for (indices[3] = 0; indices[3] < dimension.dim(3).value(); indices[3]++)
+        {
+          func(indices, dimension, channel_dim_index);
+        }
+      }
+    }
+  }
+}
+
+void cal_minmax_per_channel(const CircleNode *node, std::vector<float> &min,
+                            std::vector<float> &max)
+{
+  CircleConst *weight = loco::must_cast<CircleConst *>(get_weight(node));
+  loco::TensorShape dimension;
+  set_weight_dim(node, weight, dimension);
+
+  auto channel_dim_index = get_channel_dim_index(node);
+  auto size = dimension.dim(channel_dim_index).value();
+
+  std::vector<bool> has_min_max_value(size, false);
+  min.resize(size);
+  max.resize(size);
+
+  auto cal_minmax = [&](uint32_t *indices, loco::TensorShape &dimension,
+                        uint32_t channel_dim_index) {
+    uint32_t channel_idx = indices[channel_dim_index];
+    auto data = weight->at<loco::DataType::FLOAT32>(cal_offset(dimension, indices));
+    if (has_min_max_value[channel_idx])
+    {
+      min[channel_idx] = data < min[channel_idx] ? data : min[channel_idx];
+      max[channel_idx] = data > max[channel_idx] ? data : max[channel_idx];
+    }
+    else
+    {
+      min[channel_idx] = data;
+      max[channel_idx] = data;
+      has_min_max_value[channel_idx] = true;
+    }
+  };
+
+  iterate_per_channel(node, cal_minmax);
+}
+
+bool get_shape(const CircleNode *circle_node, std::vector<uint32_t> &shape)
+{
+  if (circle_node->shape_status() == ShapeStatus::VALID)
+  {
+    auto rank = circle_node->rank();
+    if (rank != 4)
+      return false;
+
+    shape.resize(rank);
+    for (uint32_t i = 0; i < rank; i++)
+    {
+      shape[i] = circle_node->dim(i).value();
+    }
+    return true;
+  }
+
+  return false;
+}
+
+const CircleNode *get_activation_node(const CircleNode *node)
+{
+  // only convolutions are considered currently, so just one input
+  auto preds = loco::preds(node);
+  for (const auto &pred : preds)
+  {
+    auto cur_node = loco::must_cast<const CircleNode *>(pred);
+    if (cur_node->opcode() != CircleOpcode::CIRCLECONST)
+    {
+      return cur_node;
+    }
+  }
+  return nullptr;
+}
```

Comment:
function name and body do not match. This function returns non-const predecessor. Please use clear function name (or add more comments for readability).

Reply:
Activation means input feature map?

Reply:
> Activation means input feature map?

Yes. Input feature map. Operand of considered node.

Reply:
> function name and body do not match. This function returns non-const predecessor. Please use clear function name (or add more comments for readability).

ok. I'll use `get_input_feature_map`.